opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/otlp_encoder.rs

```diff
@@ -39,50 +39,57 @@ impl OtlpEncoder {
         }
     }
 
-    pub fn encode_log_batch<'a, I>(
-        &self,
-        logs: I,
-        metadata: &str,
-    ) -> Vec<(u64, String, Vec<u8>, usize)>
+    pub fn encode_log_batch<'a, I>(&self, logs: I, metadata: &str) -> Vec<(String, Vec<u8>, usize)>
+    //(event_name, bytes, events_count)
     where
-        I: Iterator<Item = &'a opentelemetry_proto::tonic::logs::v1::LogRecord>,
+        I: IntoIterator<Item = &'a opentelemetry_proto::tonic::logs::v1::LogRecord>,
     {
         use std::collections::HashMap;
 
         let mut batches: LogBatches = HashMap::new();
 
         for log_record in logs {
-            // 1. Get schema
-            let field_specs = self.determine_fields(log_record);
-            let schema_id = Self::calculate_schema_id(&field_specs);
-            let (schema_entry, field_info) = self.get_or_create_schema(schema_id, field_specs);
-
-            // 2. Encode row
-            let row_buffer = self.write_row_data(log_record, &field_info);
+            // 1. Get schema with optimized single-pass field collection and schema ID calculation
             let event_name = if log_record.event_name.is_empty() {
                 "Log".to_string()
             } else {
                 log_record.event_name.clone()
             };
+            let (field_specs, schema_id) =
+                self.determine_fields_and_schema_id(log_record, &event_name);
+
+            let schema_entry = self.get_or_create_schema(schema_id, field_specs.clone());
+
+            // 2. Encode row
+            let row_buffer = self.write_row_data(log_record, &field_specs);
             let level = log_record.severity_number as u8;
 
-            // 3. Insert into batches - Key is (schema_id, event_name)
-            batches
-                .entry((schema_id, event_name.clone()))
-                .or_insert_with(|| (schema_entry, Vec::new()))
-                .1
-                .push((event_name, level, row_buffer));
+            // 3. Create batch key
+            let batch_key = event_name.clone();
```

Comment:
No need to clone the event_name.

```suggestion
            let batch_key = event_name;
```